File: java/change-sublayer-visibility/src/main/java/com/esri/arcgisruntime/sample/mapimagelayersublayervisibility/MainActivity.java
/* Copyright 2016 Esri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.esri.arcgisruntime.sample.mapimagelayersublayervisibility;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;

import com.esri.arcgisruntime.layers.ArcGISMapImageLayer;
import com.esri.arcgisruntime.layers.SublayerList;
import com.esri.arcgisruntime.mapping.ArcGISMap;
import com.esri.arcgisruntime.mapping.Basemap;
import com.esri.arcgisruntime.mapping.view.MapView;

public class MainActivity extends AppCompatActivity {

    private MapView mMapView;
    private ArcGISMapImageLayer mMapImageLayer;
    private SublayerList mLayers;

    // The layer on/off menu items.
    private MenuItem mCities = null;
    private MenuItem mContinent = null;
    private MenuItem mWorld = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // inflate MapView from layout
        mMapView = (MapView) findViewById(R.id.mapView);
        // create a map with the Basemap Type topographic
        ArcGISMap map = new ArcGISMap(Basemap.Type.TOPOGRAPHIC, 48.354406, -99.998267, 2);
        // create a MapImageLayer with dynamically generated map images
        mMapImageLayer = new ArcGISMapImageLayer(getResources().getString(R.string.world_cities_service));
        mMapImageLayer.setOpacity(0.5f);
        // add world cities layers as map operational layer
        map.getOperationalLayers().add(mMapImageLayer);
        // set the map to be displayed in this view
        mMapView.setMap(map);
        // get the layers from the map image layer
        mLayers = mMapImageLayer.getSublayers();

    }

    @Override
    protected void onPause(){
        super.onPause();
        mMapView.pause();
    }

    @Override
    protected void onResume(){
        super.onResume();
        mMapView.resume();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);

        // Get the sub layer switching menu items.
        mCities = menu.getItem(0);
        mContinent = menu.getItem(1);
        mWorld = menu.getItem(2);

        // set all layers on by default
        mCities.setChecked(true);
        mContinent.setChecked(true);
        mWorld.setChecked(true);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // handle menu item selection
        //if-else is used because this sample is used elsewhere as a Library module
        int itemId = item.getItemId();
        if(itemId == R.id.Cities){
            if(mLayers.get(0).isVisible() && mCities.isChecked()){
                // cities layer is on and menu item checked
                mLayers.get(0).setVisible(false);
                mCities.setChecked(false);
            }else if(!mLayers.get(0).isVisible() && !mCities.isChecked()){
                // cities layer is off and menu item unchecked
                mLayers.get(0).setVisible(true);
                mCities.setChecked(true);
            }
            return true;
        }else if(itemId == R.id.Continents){
            if(mLayers.get(1).isVisible() && mContinent.isChecked()){
                // continent layer is on and menu item checked
                mLayers.get(1).setVisible(false);
                mContinent.setChecked(false);
            }else if(!mLayers.get(1).isVisible() && !mContinent.isChecked()){
                // continent layer is off and menu item unchecked
                mLayers.get(1).setVisible(true);
                mContinent.setChecked(true);
            }
            return true;
        }else if(itemId == R.id.World){
            if(mLayers.get(2).isVisible() && mWorld.isChecked()){
                // world layer is on and menu item checked
                mLayers.get(2).setVisible(false);
                mWorld.setChecked(false);
            }else if(!mLayers.get(2).isVisible() && !mWorld.isChecked()){
                // world layer is off and menu item unchecked
                mLayers.get(2).setVisible(true);
                mWorld.setChecked(true);
            }
            return true;
        }else{
                return super.onOptionsItemSelected(item);
            }
    }

}

File: kotlin/group-layers/src/main/java/com/esri/arcgisruntime/sample/grouplayers/MainActivity.kt
/*
 * Copyright 2020 Esri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.esri.arcgisruntime.sample.grouplayers

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.recyclerview.widget.LinearLayoutManager
import com.esri.arcgisruntime.data.ServiceFeatureTable
import com.esri.arcgisruntime.layers.ArcGISSceneLayer
import com.esri.arcgisruntime.layers.FeatureLayer
import com.esri.arcgisruntime.layers.GroupLayer
import com.esri.arcgisruntime.layers.GroupVisibilityMode
import com.esri.arcgisruntime.layers.Layer
import com.esri.arcgisruntime.mapping.ArcGISScene
import com.esri.arcgisruntime.mapping.Basemap
import com.esri.arcgisruntime.mapping.LayerList
import com.esri.arcgisruntime.mapping.view.Camera
import com.google.android.material.bottomsheet.BottomSheetBehavior
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.bottom_sheet.view.*


class MainActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    // create different types of layers
    val trees =
      ArcGISSceneLayer("https://tiles.arcgis.com/tiles/P3ePLMYs2RVChkJx/arcgis/rest/services/DevA_Trees/SceneServer/layers/0").apply {
        name = "Trees"
      }
    val pathways =
      FeatureLayer(ServiceFeatureTable("https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/DevA_Pathways/FeatureServer/1")).apply {
        name = "Pathways"
      }
    val projectArea =
      FeatureLayer(ServiceFeatureTable("https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/DevelopmentProjectArea/FeatureServer/0")).apply {
        name = "Project area"
        // set the scene's viewpoint based on this layer's extent
        addDoneLoadingListener {
          sceneView.setViewpointCamera(
            Camera(
              fullExtent.center,
              700.0,
              0.0,
              60.0,
              0.0
            )
          )
        }
      }
    val buildingsA =
      ArcGISSceneLayer("https://tiles.arcgis.com/tiles/P3ePLMYs2RVChkJx/arcgis/rest/services/DevA_BuildingShells/SceneServer/layers/0").apply {
        name = "Dev A"
      }
    val buildingsB =
      ArcGISSceneLayer("https://tiles.arcgis.com/tiles/P3ePLMYs2RVChkJx/arcgis/rest/services/DevB_BuildingShells/SceneServer/layers/0").apply {
        name = "Dev B"
      }

    // create a group layer from scratch by adding the trees, pathways, and project area as children
    val projectAreaGroupLayer = GroupLayer().apply {
      name = "Project area group"
      layers.addAll(arrayOf(trees, pathways, projectArea))
    }
    // create a group layer for the buildings and set its visibility mode to exclusive
    val buildingsGroupLayer = GroupLayer().apply {
      name = "Buildings group"
      layers.addAll(arrayOf(buildingsA, buildingsB))
      visibilityMode = GroupVisibilityMode.EXCLUSIVE
    }

    // create a scene with an imagery basemap
    val scene = ArcGISScene(Basemap.createImagery()).apply {
      // add the group layer and other layers to the scene as operational layers
      operationalLayers.addAll(arrayOf(projectAreaGroupLayer, buildingsGroupLayer))
      addDoneLoadingListener {
        setupBottomSheet(operationalLayers)
      }
    }
    // set the scene to be displayed in the scene view
    sceneView.scene = scene
  }

  private fun onLayerCheckedChanged(layer: Layer, isChecked: Boolean) {
    layer.isVisible = isChecked
  }

  /** Creates a bottom sheet to display a list of group layers.
   *
   * @param layers a list of layers and group layers to be displayed on the scene
   */
  private fun setupBottomSheet(layers: LayerList) {
    // create a bottom sheet behavior from the bottom sheet view in the main layout
    val bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet).apply {
      // expand the bottom sheet, and ensure it is displayed on the screen when collapsed
      state = BottomSheetBehavior.STATE_EXPANDED
      peekHeight = bottomSheet.header.height
      // animate the arrow when the bottom sheet slides
      addBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
        override fun onSlide(bottomSheet: View, slideOffset: Float) {
          bottomSheet.header.arrowImageView.rotation = slideOffset * 180f
        }

        override fun onStateChanged(bottomSheet: View, newState: Int) {
          bottomSheet.header.arrowImageView.rotation = when (newState) {
            BottomSheetBehavior.STATE_EXPANDED -> 180f
            else -> bottomSheet.header.arrowImageView.rotation
          }
        }
      })
    }

    bottomSheet.apply {
      visibility = View.VISIBLE
      // expand or collapse the bottom sheet when the header is clicked
      header.setOnClickListener {
        bottomSheetBehavior.state = when (bottomSheetBehavior.state) {
          BottomSheetBehavior.STATE_COLLAPSED -> BottomSheetBehavior.STATE_EXPANDED
          else -> BottomSheetBehavior.STATE_COLLAPSED
        }
      }

      // initialize the recycler view with the group layers and set the callback for the checkboxes
      recyclerView.adapter = LayerListAdapter(layers) { layer: Layer, isChecked: Boolean ->
        onLayerCheckedChanged(
          layer,
          isChecked
        )
      }
      recyclerView.layoutManager = LinearLayoutManager(applicationContext)
      // rotate the arrow so it starts off in the correct rotation
      header.arrowImageView.rotation = 180f
    }

    // shrink the scene view so it is not hidden under the bottom sheet header when collapsed
    (sceneView.layoutParams as CoordinatorLayout.LayoutParams).bottomMargin =
      bottomSheet.header.height
  }

  override fun onResume() {
    super.onResume()
    sceneView.resume()
  }

  override fun onPause() {
    sceneView.pause()
    super.onPause()
  }

  override fun onDestroy() {
    sceneView.dispose()
    super.onDestroy()
  }
}

